azuread/fedauth_adal.go

```diff
@@ -0,0 +1,136 @@
+package azuread
+
+import (
+	"context"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+
+	"github.com/Azure/go-autorest/autorest/adal"
+	mssql "github.com/denisenkom/go-mssqldb"
+)
+
+type azureFedAuthConfigurer struct{}
+
+func init() {
+	mssql.SetFederatedAuthenticationConfigurer(&azureFedAuthConfigurer{})
+}
+
+// When the security token library is used, the token is obtained without input
+// from the server, so the AD endpoint and Azure SQL resource URI are provided
+// from the constants below.
+const (
+	// activeDirectoryEndpoint is the security token service URL to use when
+	// the server does not provide the URL.
+	activeDirectoryEndpoint = "https://login.microsoftonline.com/"
+
+	// azureSQLResource is the AD resource to use when the server does not
+	// provide the resource.
+	azureSQLResource = "https://database.windows.net/"
+
+	// driverClientID is the AD client ID to use when performing a username
+	// and password login.
+	driverClientID = "7f98cb04-cd1e-40df-9140-3bf7e2cea4db"
+)
+
+func retrieveToken(ctx context.Context, token *adal.ServicePrincipalToken) (string, error) {
+	err := token.RefreshWithContext(ctx)
```

Comment:
should this rather use `token.EnsureFreshWithContext(ctx)`?

Reply:
Possibly! I don't recall deciding not to use that, I think the RefreshWithContext call was part of an ADAL code sample that I just used as-is.